File: src/Manipulators/Background.php
<?php

namespace AndriesLouw\imagesweserv\Manipulators;

use AndriesLouw\imagesweserv\Manipulators\Helpers\Color;
use Jcupitt\Vips\BandFormat;
use Jcupitt\Vips\Image;

/**
 * @property string $bg
 * @property string $t
 * @property bool $isPremultiplied
 */
class Background extends BaseManipulator
{
    /**
     * Perform background image manipulation.
     *
     * @param Image $image The source image.
     *
     * @return Image The manipulated image.
     *
     * @SuppressWarnings(PHPMD.NPathComplexity)
     */
    public function run(Image $image): Image
    {
        // Skip this manipulator if:
        // - There's no bg parameter.
        // - Letterboxing is required.
        // - The image doesn't have an alpha channel.
        if (!$this->bg || $this->t === 'letterbox' || !$image->hasAlpha()) {
            return $image;
        }

        $background = new Color($this->bg);

        if ($background->isTransparent()) {
            return $image;
        }

        $backgroundRGBA = $background->toRGBA();

        if ($image->bands > 2 && $background->hasAlphaChannel()) {
            // If the image has more than two bands and the requested background color has an alpha channel;
            // alpha compositing.

            // Create a new image from a constant that matches the origin image dimensions
            $backgroundImage = $image->newFromImage([
                $backgroundRGBA[0],
                $backgroundRGBA[1],
                $backgroundRGBA[2],
                $backgroundRGBA[3]
            ]);

            // Ensure overlay is premultiplied sRGB
            $backgroundImage = $backgroundImage->premultiply();

            // Premultiply image alpha channel before background transformation
            if (!$this->isPremultiplied) {
                $image = $image->premultiply();
                $this->isPremultiplied = true;
            }

            /*
             * Alpha composite src over dst
             * Assumes alpha channels are already premultiplied and will be unpremultiplied after
             */
            $image = Image::composite(
                [$backgroundImage, $image],
                [2/*VipsBlendMode::OVER*/],
                ['premultiplied' => true]
            );
        } else {
            // If it's a 8bit-alpha channel image or the requested background color hasn't an alpha channel;
            // then flatten the alpha out of an image, replacing it with a constant background color.
            $backgroundColor = [
                $backgroundRGBA[0],
                $backgroundRGBA[1],
                $backgroundRGBA[2]
            ];

            if ($image->bands < 3) {
                // Convert sRGB to greyscale
                $backgroundColor = (0.2126 * $backgroundRGBA[0]) +
                    (0.7152 * $backgroundRGBA[1]) +
                    (0.0722 * $backgroundRGBA[2]);
            }

            // Flatten on premultiplied images causes weird results
            // so unpremultiply if we have a premultiplied image.
            if ($this->isPremultiplied) {
                // Unpremultiply image alpha and cast pixel values to integer
                $image = $image->unpremultiply()->cast(BandFormat::UCHAR);

                $this->isPremultiplied = false;
            }

            $image = $image->flatten([
                'background' => $backgroundColor
            ]);
        }


        return $image;
    }
}

File: src/weserv/manipulators/background.lua
local color = require "weserv.helpers.color"

--- Background manipulator
-- @module background
local manipulator = {}

--- Should this manipulator process the image?
-- @param args The URL query arguments.
-- @return Boolean indicating if we should process the image.
function manipulator.should_process(args)
    -- Skip this manipulator if:
    -- - There's no bg parameter.
    -- - The image doesn't have an alpha channel.
    return args.bg ~= nil and args.has_alpha
end

--- Perform background image manipulation.
-- @param image The source image.
-- @param args The URL query arguments.
-- @return The manipulated image.
function manipulator.process(image, args)
    local background = color.new(args.bg)

    -- Make sure that this manipulator is required.
    if background:is_transparent() then
        return image
    end

    local background_rgba = background:to_rgba()

    if image:bands() > 2 and background:has_alpha_channel() then
        -- If the image has more than two bands and the requested background color has an alpha channel;
        -- alpha compositing.

        -- Create a new image from a constant that matches the origin image dimensions
        local background_image = image:new_from_image(background_rgba)

        -- Ensure overlay is premultiplied sRGB
        background_image = background_image:premultiply()

        if not args.is_premultiplied then
            -- Premultiply image alpha channel before background transformation
            image = image:premultiply()
            args.is_premultiplied = true
        end

        -- Alpha composite src over dst
        -- Assumes alpha channels are already premultiplied and will be unpremultiplied after
        image = background_image:composite(image, "over", { premultiplied = true })
    else
        -- If it's a 8bit-alpha channel image or the requested background color hasn't an alpha channel;
        -- then flatten the alpha out of an image, replacing it with a constant background color.
        local background_color = {
            background_rgba[1],
            background_rgba[2],
            background_rgba[3]
        }

        if image:bands() < 3 then
            -- Convert sRGB to greyscale
            background_color = (0.2126 * background_rgba[1]) +
                    (0.7152 * background_rgba[2]) +
                    (0.0722 * background_rgba[3])
        end

        -- Flatten on premultiplied images causes weird results
        -- so unpremultiply if we have a premultiplied image.
        if args.is_premultiplied then
            -- Unpremultiply image alpha and cast pixel values to integer
            image = image:unpremultiply():cast("uchar")
            args.is_premultiplied = false
        end

        image = image:flatten({ background = background_color })

        -- The image don't have an alpha channel now
        args.has_alpha = false
    end

    return image
end

return manipulator